### ISSUE TITLE
[BUG] tuple decode assumes homogenous sequence

### ISSUE THREAD

**Issue by skeggse** [lidatong/dataclasses-json#427]
### Description

If I express a field type as `tuple[int, str]`, I would expect the two dimensions of the tuple to be decoded as each separate type. This is especially true if the type of the first dimension is an Enum, because assuming that only the first type applies to all values means decoding the second value as an Enum even though it isn't the appropriate value.

### Code snippet that reproduces the issue

```py
from dataclasses import dataclass, field
from enum import Enum

from dataclasses_json import config, dataclass_json


class ExampleEnum(Enum):
    FIRST = "first"


@dataclass_json
@dataclass
class Example:
    example_inner: tuple[ExampleEnum, str]


# This should produce no error. On the contrary, this should be perfectly valid.
Example.from_dict(dict(example_inner=("first", "second")))
```

### Describe the results you expected

No error. Instead, got:
```py
Traceback (most recent call last):
  File "/tmp/t.py", line 18, in <module>
    Example.from_dict(dict(example_inner=("first", "second")))
  File "/usr/local/lib/python3.9/site-packages/dataclasses_json/api.py", line 70, in from_dict
    return _decode_dataclass(cls, kvs, infer_missing)
  File "/usr/local/lib/python3.9/site-packages/dataclasses_json/core.py", line 210, in _decode_dataclass
    init_kwargs[field.name] = _decode_generic(field_type,
  File "/usr/local/lib/python3.9/site-packages/dataclasses_json/core.py", line 279, in _decode_generic
    res = t(xs)
  File "/usr/local/lib/python3.9/site-packages/dataclasses_json/core.py", line 339, in <genexpr>
    items = (_decode_generic(type_arg, x, infer_missing) for x in xs)
  File "/usr/local/lib/python3.9/site-packages/dataclasses_json/core.py", line 258, in _decode_generic
    res = type_(value)
  File "/usr/local/Cellar/python@3.9/3.9.16/Frameworks/Python.framework/Versions/3.9/lib/python3.9/enum.py", line 384, in __call__
    return cls.__new__(cls, value)
  File "/usr/local/Cellar/python@3.9/3.9.16/Frameworks/Python.framework/Versions/3.9/lib/python3.9/enum.py", line 702, in __new__
    raise ve_exc
ValueError: 'second' is not a valid ExampleEnum
```

### Python version you are using

3.9.16

### Environment description

irrelevant, cause is obvious: lack of supported functionality, but in this case we should see an error due to no support for tuples rather than an error due to trying to decode the enum.

**Comment by george-zubrienko** [comment]
Hi @skeggse , thanks for raising the issue. However, I'd appreciate if you avoid assertive statements when filling the issue template.
 
First, what is obvious for you, might not be obvious for others. 

Second, people that are looking for similar issues usually check environment setup (other packages etc). Unless you are using a package manager (pip isn't one), your environment might have incompatible versions of packages this library relies on.

We have a bit of a backlog here (mypy wise as you have noticed), thus I'll take a thorough look in about 2 weeks. Thanks again for contributions :)

**Comment by george-zubrienko** [comment]
@matt035343 do you think this is really an issue? The example demonstrates a behaviour that is not something you would find in json files?

**Comment by healthmatrice** [comment]
any progress about this? It is rather common to have several different types in a Tuple, though the json representation can be the same.

**Comment by healthmatrice** [comment]
actually mashmallow support this:

```python

from dataclasses import dataclass, field
from enum import Enum

from typing import Tuple


class ExampleEnum(Enum):
    FIRST = "first"


from marshmallow import Schema, fields, INCLUDE, ValidationError

class ExampleSchema(Schema):
    example_inner = fields.Tuple((fields.Enum(ExampleEnum, by_value=True), fields.Str()))

d = ExampleSchema().load(dict(example_inner=("first", "second")))
print(d)
```

It prints `{'example_inner': (<ExampleEnum.FIRST: 'first'>, 'second')}`

**Comment by george-zubrienko** [comment]
Dict is not a json. I am not convinced at all this is an issue that cannot be solved by a custom decoder.

**Comment by matt035343** [comment]
@george-zubrienko I tend to agree with @healthmatrice and @skeggse that this is a bug and it should be investigated

**Comment by Fatal1ty** [comment]
I agree that this issue should be fixed. You never know which structures a user will use and using tuples with different types is not wrong. As mentioned above, other libraries support tuples. An alternative library [mashumaro](https://github.com/Fatal1ty/mashumaro) supports this for json as well:

```python
from dataclasses import dataclass
from enum import Enum
from mashumaro.mixins.json import DataClassJSONMixin

class ExampleEnum(Enum):
    FIRST = "first"

@dataclass
class Example(DataClassJSONMixin):
    example_inner: tuple[ExampleEnum, str]

Example.from_dict({"example_inner": ("first", "second")})
# Example(example_inner=(<ExampleEnum.FIRST: 'first'>, 'second'))
Example.from_json('{"example_inner": ["first", "second"]}')
# Example(example_inner=(<ExampleEnum.FIRST: 'first'>, 'second'))
Example((ExampleEnum.FIRST, "second")).to_dict()
# {'example_inner': ['first', 'second']}
Example((ExampleEnum.FIRST, "second")).to_json()
# '{"example_inner": ["first", "second"]}'

```

**Comment by matt035343** [comment]
Hi all, I have made a PR for this issue:
https://github.com/lidatong/dataclasses-json/pull/435

**Comment by george-zubrienko** [comment]
@skeggse @healthmatrice @Fatal1ty  this will be shipped in 0.6.0 - thanks for pushing this through :)
